File: src/main/kotlin/dev/jefy/connectpro/portfolio/applicaion/dtos/GeneralInfoResponse.java
package dev.jefy.connectpro.portfolio.applicaion.dtos;

import java.util.List;

import dev.jefy.connectpro.portfolio.domain.vo.GeneralInfo;
import dev.jefy.connectpro.portfolio.domain.vo.Language;

/**
 * @author Jôph Yamba
 */
public record GeneralInfoResponse(
        String name,
        String shortDescription,
        String longDescription,
        String coverImageUrl,
        List<String> spokenLanguages
) {
    public static GeneralInfoResponse from(GeneralInfo generalInfo) {
        String imageUrl = generalInfo.coverImageUrl() ==null
                ? null
                : generalInfo.coverImageUrl().value();
        
        return new GeneralInfoResponse(
                generalInfo.name(),
                generalInfo.shortDescription(),
                generalInfo.longDescription(),
                imageUrl,
                generalInfo.spokenLanguages()
                        .stream()
                        .map(Language::value)
                        .toList()
        );
    }
}

File: src/main/kotlin/dev/jefy/connectpro/portfolio/application/dtos/GeneralInfoResponse.kt
package dev.jefy.connectpro.portfolio.application.dtos

import dev.jefy.connectpro.portfolio.domain.vo.GeneralInfo
import dev.jefy.connectpro.shared.infrastructure.file_storage.ImageUrlResolver

/**
 * @author Jôph Yamba
 */
data class GeneralInfoResponse(
    val name: String,
    val shortDescription: String,
    val longDescription: String?,
    val coverImage: String?,
    val spokenLanguages: List<String>
)
fun GeneralInfo.toResponse(resolver: ImageUrlResolver): GeneralInfoResponse = GeneralInfoResponse(
    name = this.name,
    shortDescription = this.shortDescription,
    longDescription = this.longDescription,
    coverImage = resolver.resolve(this.coverImage),
    spokenLanguages = this.spokenLanguages.map { it.value }.toList()
)
